File: 3879. Maximum Distinct Path Sum in a Binary Tree/3879. Maximum Distinct Path Sum in a Binary Tree.java
/*
 * Problem: 3879. Maximum Distinct Path Sum in a Binary Tree
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/maximum-distinct-path-sum-in-a-binary-tree/
 * Language: java
 * Date: 2026-03-26
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxSum(TreeNode root) {
        List<List<Integer>> lists = new ArrayList<>();
        List<Integer> values = new ArrayList<>();
        dfsLinks(root, -1, lists, values);
        int[][] links = new int[lists.size()][];
        int[] vals = new int[values.size()];
        for (int i = 0; i < links.length; i++) {
            List<Integer> list = lists.get(i);
            links[i] = new int[list.size()];
            for (int j = 0; j < links[i].length; j++) {
                links[i][j] = list.get(j);
            }
            vals[i] = values.get(i);
        }
        int res = -100000000;
        boolean[] seens = new boolean[2001];
        for (int i = 0; i < links.length; i++) {
            res = Math.max(res, dfs(links, vals, seens, i, -1));
        }
        return res;
    }

    public int dfsLinks(TreeNode node, int prv, List<List<Integer>> links, List<Integer> values) {
        int currentIdx = links.size();
        List<Integer> arr = new ArrayList<>();
        links.add(arr);
        values.add(node.val);
        if (prv != -1)
            arr.add(prv);
        if (node.left != null)
            arr.add(dfsLinks(node.left, currentIdx, links, values));
        if (node.right != null)
            arr.add(dfsLinks(node.right, currentIdx, links, values));
        return currentIdx;
    }

    public int dfs(int[][] links, int[] values, boolean[] seens, int i, int prv) {
        int val = values[i];
        if (seens[val + 1000])
            return -100000000;
        seens[val + 1000] = true;
        int res = 0;
        for (int next : links[i]) {
            if (next != prv) {
                res = Math.max(res, dfs(links, values, seens, next, i));
            }
        }
        seens[val + 1000] = false;
        return Math.max(res, 0) + val;
    }
}

File: 3879. Maximum Distinct Path Sum in a Binary Tree/3879. Maximum Distinct Path Sum in a Binary Tree.swift
/*
 * Problem: 3879. Maximum Distinct Path Sum in a Binary Tree
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/maximum-distinct-path-sum-in-a-binary-tree/
 * Language: swift
 * Date: 2026-03-26
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func maxSum(_ root: TreeNode?) -> Int {
        var links: [Int: [Int]] = [:]
        var values: [Int: Int] = [:]
        dfs_links(root!, -1, &links, &values)
        var res = -100000000;
        for (k, v) in links {
            var seen: Set<Int> = []
            res=max(res, dfs(&links, &values, &seen, k, -1))
        }
        return res
    }

    var idx=0;

    func dfs_links(_ node: TreeNode, _ prv: Int, _ links: inout [Int: [Int]], _ values: inout [Int: Int]) -> Int{
        var arr: [Int]=[]
        let currentIdx=idx
        idx+=1
        if prv != -1 {
            arr.append(prv)
        }
        if node.left != nil {
            arr.append(dfs_links(node.left!, currentIdx, &links, &values))
        }
        if node.right != nil {
            arr.append(dfs_links(node.right!, currentIdx, &links, &values)) 
        }
        links[currentIdx]=arr
        values[currentIdx]=node.val
        return currentIdx
    }

    func dfs(_ links: inout [Int: [Int]], _ values: inout [Int: Int], _ seen: inout Set<Int>, _ i: Int, _ prv: Int) -> Int{
        var res = 0
        let link=links[i]!
        let val=values[i]!
        if seen.contains(val) {
            return -100000000
        }
        seen.insert(val)
        for next in link {
            if next != prv {
                res=max(res, dfs(&links, &values, &seen, next, i))
            }
        }
        seen.remove(val)
        return max(res, 0)+val
    }
}
